core/trino-main/src/main/java/io/trino/operator/InterpretedHashGenerator.java

```diff
@@ -45,26 +48,45 @@ public InterpretedHashGenerator(List<Type> hashChannelTypes, List<Integer> hashC
 
     public InterpretedHashGenerator(List<Type> hashChannelTypes, int[] hashChannels, BlockTypeOperators blockTypeOperators)
     {
-        this.hashChannels = requireNonNull(hashChannels, "hashChannels is null");
+        this(hashChannelTypes, requireNonNull(hashChannels, "hashChannels is null"), blockTypeOperators, false);
+    }
+
+    private InterpretedHashGenerator(List<Type> hashChannelTypes, @Nullable int[] hashChannels, BlockTypeOperators blockTypeOperators, boolean positional)
+    {
         this.hashChannelTypes = ImmutableList.copyOf(requireNonNull(hashChannelTypes, "hashChannelTypes is null"));
-        checkArgument(hashChannelTypes.size() == hashChannels.length);
-        this.hashCodeOperators = hashChannelTypes.stream()
-                .map(blockTypeOperators::getHashCodeOperator)
-                .collect(toImmutableList());
+        this.hashCodeOperators = createHashCodeOperators(hashChannelTypes, blockTypeOperators);
+        checkArgument(hashCodeOperators.length == hashChannelTypes.size());
+        if (positional) {
+            checkArgument(hashChannels == null, "hashChannels must be null");
+            this.hashChannels = null;
+        }
+        else {
+            requireNonNull(hashChannels, "hashChannels is null");
+            checkArgument(hashChannels.length == hashCodeOperators.length);
+            // simple positional indices are converted to null
+            this.hashChannels = isPositionalChannels(hashChannels) ? null : hashChannels;
+        }
     }
 
     @Override
     public long hashPosition(int position, Page page)
     {
-        return hashPosition(position, page::getBlock);
+        // Note: this code is duplicated for performance but must logically match hashPosition(position, IntFunction<Block> blockProvider)
+        long result = HashGenerationOptimizer.INITIAL_HASH_VALUE;
+        for (int i = 0; i < hashCodeOperators.length; i++) {
+            Block block = page.getBlock(hashChannels == null ? i : hashChannels[i]);
```

Comment:
Would it make sense to move that `if` outside of loop? What would be perf difference?

Reply:
Do you mean "what if we manually unswitched the loop" (ie: what if we pushed to loop implementations underneath an `if (hashChannels == null)`)?

I suppose we could, but then we'd have 4 implementations of _almost_ the same loop body. Logically, it should be the same since we're relying on C2 to essentially do exactly that for us, but the difference would be that overall code size of this method would be larger and that might affect whether this method would get inlined elsewhere.